modules/PhenoDigm.py

```diff
@@ -0,0 +1,362 @@
+            # Rename the disease data columns.
+            .withColumnRenamed('disease_id', 'diseaseFromSourceId')
+            .withColumnRenamed('disease_term', 'diseaseFromSource')
+            # Add constant value columns.
+            .withColumn('datasourceId', pf.lit('phenodigm'))
+            .withColumn('datatypeId', pf.lit('animal_model'))
+
+            # Ensure stable column order.
+            .select('biologicalModelAllelicComposition', 'biologicalModelGeneticBackground', 'biologicalModelId',
+                    'datasourceId', 'datatypeId', 'diseaseFromSource', 'diseaseFromSourceId',
+                    'diseaseModelAssociatedHumanPhenotypes', 'diseaseModelAssociatedModelPhenotypes', 'resourceScore',
+                    'targetFromSourceId', 'targetInModel')
+        )
+
+    def write_evidence_strings(self, evidence_strings_filename):
+        """Dump the Spark evidence dataframe into a temporary directory as separate JSON chunks. Collect and combine
+        them to obtain the final output file. The order of the evidence strings is not maintained, and they are returned
+        in random order as collected by Spark."""
+        with tempfile.TemporaryDirectory() as tmp_dir_name, open(evidence_strings_filename, 'wb') as outfile:
+            (
+                self.evidence.write
+                .format('json').mode('overwrite').option('compression', 'org.apache.hadoop.io.compress.GzipCodec')
+                .save(tmp_dir_name)
+            )
+            for json_chunk_filename in [f for f in os.listdir(tmp_dir_name) if f.endswith('.json.gz')]:
+                with open(os.path.join(tmp_dir_name, json_chunk_filename), 'rb') as json_chunk:
+                    shutil.copyfileobj(json_chunk, outfile)
+
+    def process_all(self, output, score_cutoff, use_cached):
+        if not use_cached:
+            self.logger.info('Update the HGNC/MGI/SOLR cache.')
+            self.update_cache()
+
+        self.logger.info('Load gene mappings and SOLR data from local cache.')
+        self.load_data_from_cache()
+
+        self.logger.info('Build the evidence strings.')
+        self.generate_phenodigm_evidence_strings(score_cutoff)
+
+        self.logger.info('Collect and write the evidence strings.')
+        self.write_evidence_strings(output)
+
+
+def main(cache_dir, output, score_cutoff, use_cached=False, log_file=None):
+    # Initialize the logger based on the provided log file. If no log file is specified, logs are written to STDERR.
+    logging_config = {
+        'level': logging.INFO,
+        'format': '%(asctime)s %(levelname)s %(module)s - %(funcName)s: %(message)s',
+        'datefmt': '%Y-%m-%d %H:%M:%S',
+    }
+    if log_file:
+        logging_config['filename'] = log_file
+    logging.basicConfig(**logging_config)
+
+    # Process the data.
+    PhenoDigm(logging, cache_dir).process_all(output, score_cutoff, use_cached)
+
+
+if __name__ == '__main__':
```

Comment:
I personally think it is a better practice to include this logic within main and in this if block just call the main() function.

Reply:
Ah, this is where I completely disagree, if you don't mind. To me, the entire point of having the `if __name__ == '__main__'` block is to be able to import the functions (including `main()` if you want to) from the module into another one.

If we define a parser inside `main()` and then try to import it, it will break things heavily, because the command line arguments for the parser won't be there anymore.